### DISCUSSION TITLE
Reading custom fields (Graphql)?

### DISCUSSION THREAD

**Discussion by slanelb** [swellstores/community#152] [Help] [answered]
Hello, is it possible to read custom fields within the product model using Graphql SDK or Swell js or must I use the Node library?

**Comment by rodrigoddalmeida** [comment, marked as answer]
Hi @slanelb, custom fields added to the `content` namespace are available by default in our Frontend APIs. If you have a custom field out of the `content` namespace, you need to explicitly add it to your public key's permissions.

This article might help you achieve that: https://developers.swell.is/guides/opening-content-permissions

**Reply by slanelb** [reply]
Excellent! Thank you!

**Reply by steph-query** [reply]
Can you read a custom model directly via the API? I have a mode I've called `test` inside the `content` namespace. Yet when I run the following GraphQL query, I am not able to query my `test` model:

![image](https://github.com/swellstores/community/assets/8635557/ddb0ad75-3ecd-48ad-9e88-8a1f9b0172f0)


```
{
      contentTest {
        name
      }
 }
```

Am I missing something @rodrigoddalmeida ?

**Reply by logeshswell** [reply]
Yes, you can fetch the custom models via GraphQL as follows:

```graphql
query {
  contentTests {
    results {
      name
    }
  }
}
```

If this doesn't work, please help me with your store link.

**Reply by steph-query** [reply]
Thanks for looking into this. My store link is somm-demand.swell.store. 

I wasn't expecting it to pluralize the model name, but I'm still seeing the the same error: `Cannot query field "contentTests" on type "Query". Did you mean "contentPages"?`

```
const response = await makeRequest('/graphql', 'POST', {
    query: `query {
      contentTests {
        results {
          name
        }
      }
    }`,
  })
```

This is copied verbatim from the [SquareOne](https://github.com/CommitStudio/square-one-swell/blob/a818c7990ee352d55edb9972a9da82230cf11020/app/_lib/SwellAPI.ts#L31) template, and works with non-custom models. While it would be great to get this example working, my real aim here is to be able to traverse these relationships through the products. For example, I want a user to be able to see wines they have purchased, including fields like `country` and `region`, for which I created custom models and added fields linking them to products. 

Ideally I can make a query like this:
```
orders {
  results {
    items {
      product {
        name
        sku
        country {
          name
        }
        region {
          name
        }
        ...
      }
    }
  }
}
```

I created my custom models as data models, not content models, which may have been a mistake, but it was the only way I could get those fields to show up in the admin UI when creating products. Sounds like this is achievable, so I would be grateful for your guidance, as it would be a drag to have to use the SDK and multiple round trips to get all the data I need.

**Reply by logeshswell** [reply]
There's a bug found in identifying the env from the public key. So, as an immediate workaround, can you pass "swell-env": "test" in headers, so that you can fetch the expected results using the same query.

<img width="1512" alt="Screenshot 2023-10-24 at 1 16 09 PM" src="https://github.com/swellstores/community/assets/100766131/7c472a6f-9996-48c5-bd80-a72dc561d8c1">

I will keep you posted if I get some updates on the fix.

**Reply by steph-query** [reply]
Thank you for clarifying that. I am able to fetch this model directly when I update my headers. However, this query still fails:

```
{
  orders {
    results {
      items {
        product {
          id 
          test {
            name
          }
        }
      }
    }
  }
}`
```
I get the error `Cannot query field "test" on type "SwellProduct"`. I have also tried a number of ways, like nesting under a content attribute, and `contentTest`, etc. with no luck.
```
{
...
  product {
    id
    content {
      test {
        name
      }
    }
  }
}
```
I have added a field to my product model called `test` and pointed it at this collection:
<img width="576" alt="image" src="https://github.com/swellstores/community/assets/8635557/da2b3ad8-0cbc-448d-8bbf-82b2edbbbe8c">

Is there any way to fetch these kinds of content records in a query like this?

**Reply by ericingram** [reply]
Looks like it needs the content property in the query (not tested, just saying)
```
{
  orders {
    results {
      items {
        product {
          id 
          content {
            test {
              name
            }
          }
        }
      }
    }
  }
}
```

**Reply by steph-query** [reply]
Thanks @ericingram , I did give that a try and got this error:

`Cannot query field "test" on type "SwellProductContent".`

**Reply by logeshswell** [reply]
[Eric's answer](https://github.com/orgs/swellstores/discussions/152#discussioncomment-7384246) is the expected one, but it looks like a gap that we missed to support dynamic lookup fields.

As of now, you can query only `testId` as follows:
```graphql
{
  orders {
    results {
      items {
        product {
          id 
          content {
            testId
          }
        }
      }
    }
  }
}
```

I will keep you posted when we fix this gap. Thanks for bringing it up!

**Reply by steph-query** [reply]
Thanks @logeshswell! I have a lot of models that I did not create in the content namespace when i first created them (things like `countries`, `regions`, etc. for a taxonomy of geographic wine data that I attached to my products. 

I'll stay tuned for this update, but my last question is: Will I/Should I be able to access these fields through the GraphQL API?

For example, when I fetch a product in the dashboard console, I see all the links, including the `testId`:
```json
{
    "active": true,
    "glassware": "64f3a31a8928410012378661",
    "alcohol": "13%",
    "varieties": "64f39eb9990c1f00123371c1",
    "vineyard": "Vista Luna Vineyard",
    "size": "64f395a5e5a474001109619f",
    "price": 21,
    "description": "Ferdinand is a small winery focusing on Spanish varietals grown in California. The flagship wine is their Albari&ntilde;o, which comes from the Luna Vista vineyard in the Borden Ranch subzone of the Lodi AVA. Albari&ntilde;o offers an interesting combination of floral aromatics and crisp white fruit, and this one spends 8 months on the lees in used oak barrels, giving it a lush texture to balance its mouthwatering acidity.",
    "name": "Ferdinand, California Albariño 'Vista Luna Vineyard 2019 750ml",
    "sku": "prod_0egY5eMRrM53Qn",
    "slug": "ferdinand-california-albario-vista-luna-vineyard-2019-750ml",
    "producer_id": "64f39bfaa5c9ea00110172b5",
    "wine_category_id": "64f3a05e36212500129b43c1",
    "country_id": "64f395cbd66b1f0012720fea",
    ...
    "appellation_id": "64f3998921b14f0012579270",
    "region_id": "64f39604d83b3f00126f169a",
    "fermentation_id": "64f3a0676465720012c3ee45",
    "aging_id": "64f3a0676465720012c3ee45",
    "wine_type_id": "64f39de38bcca80011780af9",
    "date_updated": "2023-10-25T22:43:26.376Z",
    "test_id": "6536cbdae1c822001216c74d",
    "content": {
      "wine_type_id": null
    },
    "id": "64f3b6c8153a160012eb6b13"
}
```
However, when I run the query you shared (adding product name to the requested fields), I get this response:

```json
{
    "product": {
        "id": "64f3b6c8153a160012eb6b13",
        "name": "Ferdinand, California Albariño 'Vista Luna Vineyard 2019 750ml",
        "content": null
    }
}
```

When I ask GraphQL for products at the top level (instead of getting to them via products) and try to traverse these relationships, I get the shape back, but no values.

```json
{
    "id": "64f3b6c8153a160012eb6b13",
    "name": "Ferdinand, California Albariño 'Vista Luna Vineyard 2019 750ml",
    "content": {
        "testId": null,
        "wineTypeId": null
    }
}
```

I'm having trouble making sense of all the different response shapes and values, and how the behavior is meant to change with content models vs custom models in their own namespace.

**Reply by logeshswell** [reply]
> Will I/Should I be able to access these fields through the GraphQL API?

You can retrieve all the fields present under `content` namespace via GraphQL API.

**Reply by logeshswell** [reply]
Please refer to the **DOCS** section inside the GraphQL playground to understand the possible response schema:
https://somm-demand.swell.store/playground

**Reply by steph-query** [reply]
Thanks, what I've realized is that switching a field to move it under the content namespace clears it from the existing products! So if I had selected a country, for example, that link would disappear when I changed the country field to be `content.country`. It is unfortunate that I will have to re-connect all of my products to their related countries/regions/etc.  when I change the namespace on the field- is this intended?

That said, getting country ID from GraphQL does seem to work without actually dropping the custom countries model I made, meaning I don't _think_ I have to recreate them all as content models, I only need to change the field on products that links to the custom model (at least to get the country ID from graphQL). 

I'll be eagerly awaiting the updates that will allow me to use GraphQL to fetch actual attributes on the custom content models via the product fields, so please do let me know when that is available.  Hopefully I don't have to go and re-create/re-link everything.

**Reply by logeshswell** [reply]
> when I change the namespace on the field- is this intended?

Yes, you need to migrate the values before the migration. I will keep you posted when we fix this gap.
